R/load_forecasts_zoltar.R

```diff
@@ -0,0 +1,100 @@
+#' @param verbose a boolean for printing messages on zoltar job status. Default to TRUE.
+#'
+#' @return data frame with columns model, forecast_date, location, horizon, 
+#' temporal_resolution, target_variable, target_end_date, type, quantile, value,
+#' location_name, population, geo_type, geo_value, abbreviation
+#' 
+#' @export
+load_forecasts_zoltar <- function (
+  models = NULL,
+  forecast_dates = NULL,
+  locations = NULL,
+  types = NULL,
+  targets = NULL,
+  as_of = NULL,
+  verbose = TRUE) {
+  
+  # set up Zoltar connection
+  zoltar_connection <- zoltr::new_connection()
+  if(Sys.getenv("Z_USERNAME") == "" | Sys.getenv("Z_PASSWORD") == "") {
+    zoltr::zoltar_authenticate(zoltar_connection, "zoltar_demo","Dq65&aP0nIlG")
+  } else {
+    zoltr::zoltar_authenticate(zoltar_connection, Sys.getenv("Z_USERNAME"),Sys.getenv("Z_PASSWORD"))
+  }
+  
+  if (!is.null(forecast_dates)){
+    # construct Zoltar project url
+    the_projects <- zoltr::projects(zoltar_connection)
+    project_url <- the_projects[the_projects$name == "COVID-19 Forecasts", "url"]
+    
+    # get all valid timezeros in project
+    all_valid_timezeros <- zoltr::timezeros(zoltar_connection = zoltar_connection,
+                                            project_url = project_url)$timezero_date
+    
+    # take intersection of forecast_dates and all_valid_timezeros
+    valid_forecast_dates <- intersect(as.character(forecast_dates), 
+                                      as.character(all_valid_timezeros))
+    if (length(valid_forecast_dates) == 0) {
+      stop("Error in load_forecasts: All forecast_dates are invalid.")
+    }
+  } 
+  
+  forecasts <- zoltr::do_zoltar_query(zoltar_connection = zoltar_connection,
+                                      project_url = project_url,
+                                      query_type = "forecasts",
+                                      units = locations, 
+                                      timezeros = valid_forecast_dates,
+                                      models = models,
+                                      targets = targets,
+                                      types = types,
+                                      verbose = verbose,
+                                      as_of = as_of)
+  if (nrow(forecasts) == 0){
+    warning("Warning in do_zotar_query: Forecasts are not available.\n Please check your parameters.")
+    # convert value column to double and select columns
+    forecasts <- forecasts %>%
+      dplyr::mutate(value = as.double(value)) %>%
+      tidyr::separate(target, into=c("horizon","temporal_resolution","ahead",
+                                     "target_variable"),
+                      remove = FALSE, extra = "merge") %>%
+      dplyr::rename(location = unit, forecast_date = timezero, type = class) %>%
+      dplyr::select(model, forecast_date, location, horizon, temporal_resolution,
+                    target_variable, type, quantile, value)
```

Comment:
How much of this is necessary if we got no rows back?  Can we clarify and simplify?

Reply:
I simplified this part a bit to include fewer columns. The default data format from Zoltar query has some columns that we are not using now. This part is just cleaning up the data frame when the user is trying to bind results from load_latest_forecasts().